framework/web/UrlRule.php

```diff
@@ -256,6 +355,25 @@ public function parseRequest($manager, $request)
         }
         $matches = $this->substitutePlaceholderNames($matches);
 
+        // if routing succeeded (some rule matched), check whether pathInfo
+        //  was changed during canonization. If so, take the required action.
+        if ($pathInfoCanonized != $pathInfoOrig) {
+            switch ($this->canonize['action']) {
+                case 'redirect':
+                    $url = '/' . $pathInfoCanonized
+                        . (Yii::$app->request->queryString ? '?' . Yii::$app->request->queryString : '');
+                    die('redirect ' . $url);
+                    Yii::$app->getResponse()->redirect($url)->send();
+                    break;
```

Comment:
Of course, but since canonization is configured manually, developer should take the responsibility of adding redirect action to the controller.

We can add an option (i.e. 'force-redirect' => true) to just redirect without returning a redirection route (as canonizer currently behaves). But as forced redirect can break existing code (if someone uses parseRequest() to just get route info without routing indeed), this should not be the default, I think.